File: apps/web/src/lib/watch_payload_fixture.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { parseRunSource, type RunSource } from './types';

// Cross-platform contract test for the canonical watch-run payload.
//
// The fixture at `fixtures/watch_run_payload.json` (repo root) is shared
// with the Flutter test (apps/mobile_android + mobile_ios under
// test/watch_payload_fixture_test.dart) and the Wear OS test
// (apps/watch_wear under WatchRunPayloadFixtureTest.kt). All three
// platforms read the same file and must agree on the shape. If you
// change a field, update both other tests in the same commit.
//
// Web doesn't decode watch payloads (the watch path goes through Dart on
// iOS or directly to Supabase from Wear OS). What web does is read the
// resulting rows. So the web side of the contract is structural: the
// source must parse, the metadata keys web reads (activity_type,
// avg_bpm) must be present, and the laps shape must match the registry.

const __dirname = dirname(fileURLToPath(import.meta.url));
const FIXTURE_PATH = join(__dirname, '..', '..', '..', '..', 'fixtures', 'watch_run_payload.json');

type Fixture = {
	payload: { source: string; activity_type: string };
	expectedRow: { source: string; duration_s: number; distance_m: number };
	expectedMetadata: {
		activity_type: string;
		avg_bpm: number;
		laps: Array<{
			index: number;
			start_offset_s: number;
			distance_m: number;
			duration_s: number;
		}>;
	};
	expectedTrackCount: number;
	expectedFirstPointBpm: number;
};

const fixture = JSON.parse(readFileSync(FIXTURE_PATH, 'utf-8')) as Fixture;

test('fixture source parses to a valid RunSource', () => {
	const parsed: RunSource = parseRunSource(fixture.expectedRow.source);
	assert.equal(parsed, fixture.expectedRow.source);
	assert.equal(parsed, 'watch');
});

test('fixture metadata.activity_type is one of the registered values', () => {
	// Match docs/metadata.md § activity_type.
	const allowed = new Set(['run', 'walk', 'hike', 'cycle']);
	assert.ok(
		allowed.has(fixture.expectedMetadata.activity_type),
		`activity_type=${fixture.expectedMetadata.activity_type} not in registered set`,
	);
});

test('fixture metadata.avg_bpm is a positive number', () => {
	assert.equal(typeof fixture.expectedMetadata.avg_bpm, 'number');
	assert.ok(fixture.expectedMetadata.avg_bpm > 0);
});

test('fixture laps use canonical per-lap-delta shape', () => {
	// Per docs/metadata.md § laps: index (int), start_offset_s (int),
	// distance_m (number), duration_s (int). start_offset_s of the first
	// lap is 0 (cumulative-BEFORE), and subsequent laps' start_offset_s
	// equals the running total of prior duration_s values.
	const laps = fixture.expectedMetadata.laps;
	assert.ok(laps.length >= 2, 'fixture must have ≥ 2 laps to test offset accumulation');
	assert.equal(laps[0].start_offset_s, 0, 'first lap must start at offset 0');
	let runningOffset = 0;
	for (const lap of laps) {
		assert.equal(typeof lap.index, 'number');
		assert.equal(typeof lap.start_offset_s, 'number');
		assert.equal(typeof lap.distance_m, 'number');
		assert.equal(typeof lap.duration_s, 'number');
		assert.equal(
			lap.start_offset_s,
			runningOffset,
			`lap ${lap.index}: start_offset_s should equal cumulative duration of prior laps`,
		);
		runningOffset += lap.duration_s;
	}
});

test('fixture row source matches payload source', () => {
	// If these drift, the watch ingestion pipeline misclassifies runs.
	assert.equal(fixture.payload.source, fixture.expectedRow.source);
});

test('fixture activity_type matches between payload and metadata', () => {
	assert.equal(
		fixture.payload.activity_type,
		fixture.expectedMetadata.activity_type,
		'payload activity_type must equal what lands in metadata',
	);
});

File: apps/mobile_ios/test/watch_payload_fixture_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import '../lib/watch_ingest_queue.dart';

/// Cross-platform contract test for the canonical watch-run payload.
///
/// The fixture at `fixtures/watch_run_payload.json` (repo root) is shared
/// with the Wear OS test (`apps/watch_wear/.../WatchRunPayloadFixtureTest.kt`)
/// and the web test (`apps/web/src/lib/watch_payload_fixture.test.ts`).
/// All three platforms decode the same file and must agree on the shape.
/// If you change a field here, update both other tests in the same commit.
///
/// This is the guardrail for the class of bugs the April 2026 audit found:
/// laps shape divergence, missing per-point bpm, missing activity_type.
void main() {
  group('canonical watch-run payload fixture', () {
    late Map<String, dynamic> fixture;
    late Map<String, dynamic> payload;
    late Map<String, dynamic> expectedRow;
    late Map<String, dynamic> expectedMetadata;

    setUpAll(() {
      // The test runner cwd is the package dir
      // (apps/mobile_android/), so repo root is two parents up.
      final file = File('../../fixtures/watch_run_payload.json');
      fixture = jsonDecode(file.readAsStringSync()) as Map<String, dynamic>;
      payload = fixture['payload'] as Map<String, dynamic>;
      expectedRow = fixture['expectedRow'] as Map<String, dynamic>;
      expectedMetadata = fixture['expectedMetadata'] as Map<String, dynamic>;
    });

    test('row fields match expectedRow', () {
      final run = runFromWatchPayload(payload);
      expect(run.source.name, expectedRow['source']);
      expect(run.duration.inSeconds, expectedRow['duration_s']);
      expect(run.distanceMetres, expectedRow['distance_m']);
    });

    test('metadata.activity_type round-trips', () {
      final run = runFromWatchPayload(payload);
      expect(run.metadata?['activity_type'], expectedMetadata['activity_type']);
    });

    test('metadata.avg_bpm round-trips as a number', () {
      final run = runFromWatchPayload(payload);
      // avg_bpm is stored as double (the existing decoder casts to double).
      expect(run.metadata?['avg_bpm'], expectedMetadata['avg_bpm']);
    });

    test('metadata.laps preserves canonical per-lap shape', () {
      final run = runFromWatchPayload(payload);
      final laps = (run.metadata?['laps'] as List).cast<Map>();
      final expectedLaps = (expectedMetadata['laps'] as List).cast<Map>();
      expect(laps.length, expectedLaps.length);
      for (var i = 0; i < laps.length; i++) {
        expect(laps[i]['index'], expectedLaps[i]['index']);
        expect(laps[i]['start_offset_s'], expectedLaps[i]['start_offset_s']);
        expect(laps[i]['distance_m'], expectedLaps[i]['distance_m']);
        expect(laps[i]['duration_s'], expectedLaps[i]['duration_s']);
      }
    });

    test('track waypoint count + per-point bpm round-trip', () {
      final run = runFromWatchPayload(payload);
      expect(run.track.length, fixture['expectedTrackCount']);
      expect(run.track.first.bpm, fixture['expectedFirstPointBpm']);
      // Elevation + timestamp also round-trip on the first point.
      expect(run.track.first.elevationMetres, isNotNull);
      expect(run.track.first.timestamp, isNotNull);
    });
  });
}
